aptos-move/mvhashmap/src/unsync_map.rs

```diff
@@ -0,0 +1,145 @@
+/// UnsyncMap utilizes RefCell to provide access with interior mutability.
+pub struct UnsyncMap<K: ModulePath, V: TransactionWrite, X: Executable> {
+    data_map: RefCell<BTreeMap<K, Entry<V, X>>>,
+    executable_cache: RefCell<HashMap<HashValue, X>>,
+    executable_bytes: RefCell<usize>,
+}
+
+impl<K: ModulePath + Hash + Clone + Eq + Ord, V: TransactionWrite + Clone, X: Executable> Default
+    for UnsyncMap<K, V, X>
+{
+    fn default() -> Self {
+        Self {
+            data_map: RefCell::new(BTreeMap::new()),
+            executable_cache: RefCell::new(HashMap::new()),
+            executable_bytes: RefCell::new(0),
+        }
+    }
+}
+
+impl<K: ModulePath + Hash + Clone + Eq + Ord, V: TransactionWrite + Clone, X: Executable>
+    UnsyncMap<K, V, X>
+{
+    pub fn fetch_data(&self, key: &K) -> Option<V> {
+        self.data_map
+            .borrow()
+            .get(key)
+            .and_then(|entry| entry.data.clone())
+    }
+
+    pub fn fetch_code(&self, key: &K) -> Option<MVCodeOutput<V, X>> {
+        use MVCodeOutput::*;
+        debug_assert!(key.module_path().is_some());
+
+        self.data_map
+            .borrow_mut()
+            .get_mut(key)
+            .and_then(|entry| match &entry.data {
+                Some(module) => {
+                    let hash = entry.hash.get_or_insert(module_hash(module));
+
+                    Some(self.executable_cache.borrow().get(hash).map_or_else(
+                        || Module((module.clone(), *hash)),
+                        |x| Executable((x.clone(), ExecutableDescriptor::Published(*hash))),
+                    ))
+                },
+                None => entry
+                    .maybe_base_executable
+                    .as_ref()
+                    .map(|x| Executable((x.clone(), ExecutableDescriptor::Storage))),
+            })
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        self.data_map
+            .borrow_mut()
+            .insert(key, Entry::from_data(value));
+    }
+
+    /// We return false if the executable was already stored, as this isn't supposed to happen
+    /// during sequential execution (and the caller may choose to e.g. log a message).
+    pub fn store_executable(
+        &self,
+        key: &K,
+        descriptor: ExecutableDescriptor,
+        executable: X,
+    ) -> bool {
+        let size = executable.size_bytes();
+        let ret = match descriptor {
+            ExecutableDescriptor::Published(hash) => self
+                .executable_cache
+                .borrow_mut()
+                .insert(hash, executable)
+                .is_none(),
+            ExecutableDescriptor::Storage => match self.data_map.borrow_mut().entry(key.clone()) {
+                Occupied(mut entry) => entry
+                    .get_mut()
+                    .maybe_base_executable
+                    .replace(executable)
+                    .is_none(),
```

Comment:
What is the point of storing it in this case?
If it is Occupied, it means that the entry was created from data, right?
In this case, we will return from the executable cache anyway.

Actually, is not this unreachable?

Reply:
it was just confusing design, in sequential execution i had "simplified" and assumed ExecutableCache isn't going to make a difference (at least for now), so didn't even assume that exists as a baseline. But then during block execution we might fetch some module at storage level and prepare executable for it, then ask to have it stored. It should have been stored at least during the block execution, and as I didn't have the base executables in the view, and MVCodeMap dealt with it, I ended up caching them here. This is of course horrible.

With the suggestion to move base executables to the view, this is indeed no longer needed here.